File: solutions/java/src/main/java/com/epi/SmallestSubarrayCoveringSet.java
// Copyright (c) 2015 Elements of Programming Interviews. All rights reserved.

package com.epi;

import java.util.*;

class SmallestSubarrayCoveringSet {
  public static String randString(int len) {
    StringBuilder sb = new StringBuilder();
    Random gen = new Random();
    while (len-- != 0) {
      sb.append((char)(gen.nextInt('z' + 1 - 'a') + 'a'));
    }
    return sb.toString();
  }

  // @include
  // Represent subarray by starting and ending indices, inclusive.
  private static class Subarray {
    public Integer start;
    public Integer end;

    public Subarray(Integer start, Integer end) {
      this.start = start;
      this.end = end;
    }
  }

  public static Subarray findSmallestSubarrayCoveringSet(String[] paragraph,
                                                         Set<String> keywords) {
    Map<String, Integer> keywordsToCount = new HashMap<>();
    int left = 0, right = 0;
    Subarray result = new Subarray(-1, -1);
    while (right < paragraph.length) {
      // Keeps advancing right until it reaches end or keywordsToCount has
      // all keywords.
      while (right < paragraph.length &&
             keywordsToCount.size() < keywords.size()) {
        if (keywords.contains(paragraph[right])) {
          keywordsToCount.put(paragraph[right],
                              keywordsToCount.containsKey(paragraph[right])
                                  ? keywordsToCount.get(paragraph[right]) + 1
                                  : 1);
        }
        ++right;
      }

      // Found all keywords, update the smallest subarray containing all keywords.
      if (keywordsToCount.size() == keywords.size() &&
          ((result.start == -1 && result.end == -1) ||
           right - 1 - left < result.end - result.start)) {
        result.start = left;
        result.end = right - 1;
      }

      // Keeps advancing left until it reaches end or keywordsToCount does not
      // have all keywords.
      while (left < right && keywordsToCount.size() == keywords.size()) {
        if (keywords.contains(paragraph[left])) {
          int keywordCount = keywordsToCount.get(paragraph[left]);
          keywordsToCount.put(paragraph[left], --keywordCount);
          if (keywordCount == 0) {
            keywordsToCount.remove(paragraph[left]);
            if ((result.start == -1 && result.end == -1) ||
                right - 1 - left < result.end - result.start) {
              result.start = left;
              result.end = right - 1;
            }
          }
        }
        ++left;
      }
    }
    return result;
  }
  // @exclude

  // O(n^2) solution
  public static int checkAns(String[] A, String[] Q) {
    Set<String> dict = new HashSet<>(Arrays.asList(Q));

    Subarray ans = new Subarray(0, A.length - 1);
    for (int l = 0; l < A.length; ++l) {
      Map<String, Integer> count = new HashMap<>();
      for (int r = l; r < A.length && r - l < ans.end - ans.start; ++r) {
        if (dict.contains(A[r])) {
          count.put(A[r], count.containsKey(A[r]) ? count.get(A[r]) + 1 : 1);
        }
        if (count.size() == Q.length) {
          if (r - l < ans.end - ans.start) {
            ans.start = l;
            ans.end = r;
          }
          break;
        }
      }
      count.clear();
    }

    return ans.end - ans.start;
  }

  public static void main(String[] args) {
    Random gen = new Random();
    for (int times = 0; times < 1000; ++times) {
      int n;
      if (args.length == 1) {
        n = Integer.parseInt(args[0]);
      } else {
        n = gen.nextInt(10000) + 1;
      }
      String[] A = new String[n];
      for (int i = 0; i < n; ++i) {
        A[i] = randString(gen.nextInt(10) + 1);
      }
      /*
       * for (int i = 0; i < A.size(); ++i) { cout << A[i] << ' '; } cout <<
       * endl;
       */
      Set<String> dict = new HashSet<>(Arrays.asList(A));
      int m = gen.nextInt(dict.size()) + 1;
      String[] Q = new String[m];
      int idx = 0;
      for (String aDict : dict) {
        Q[idx++] = aDict;
        if (--m == 0) {
          break;
        }
      }

      /*
       * for (int i = 0; i < Q.size(); ++i) { cout << Q[i] << ' '; } cout <<
       * endl;
       */

      dict = new HashSet<>(Arrays.asList(Q));
      Subarray res = findSmallestSubarrayCoveringSet(A, dict);
      System.out.println(res.start + ", " + res.end);
      dict.clear();
      for (String aQ1 : Q) {
        dict.add(aQ1);
      }
      for (int i = res.start; i <= res.end; ++i) {
        if (dict.contains(A[i])) {
          dict.remove(A[i]);
        }
      }
      assert(dict.isEmpty());
      Integer[] res2Array = SmallestSubarrayCoveringSetStream
                                .findSmallestSubarrayCoveringSubsetAsArray(A, Q);
      Subarray res2 = new Subarray(res2Array[0], res2Array[1]);

      System.out.println(res2.start + ", " + res2.end);
      dict.clear();
      for (String aQ : Q) {
        dict.add(aQ);
      }
      for (int i = res.start; i <= res.end; ++i) {
        if (dict.contains(A[i])) {
          dict.remove(A[i]);
        }
      }
      assert(dict.isEmpty());
      assert(res.end - res.start == res2.end - res2.start);
      assert(res.end - res.start == checkAns(A, Q));
    }
  }
}

File: static/solutions/cpp/Smallest_subarray_covering_set.cc
// Copyright (c) 2015 Elements of Programming Interviews. All rights reserved.

#include <cassert>
#include <iostream>
#include <random>
#include <sstream>
#include <string>
#include <unordered_map>
#include <unordered_set>
#include <utility>
#include <vector>

#include "./Smallest_subarray_covering_set.h"
#include "./Smallest_subarray_covering_set_stream.h"

using std::cout;
using std::default_random_engine;
using std::endl;
using std::istringstream;
using std::random_device;
using std::uniform_int_distribution;
using std::unordered_map;
using std::unordered_set;
using std::vector;

string RandString(int len) {
  string ret;
  default_random_engine gen((random_device())());
  while (len--) {
    uniform_int_distribution<int> dis('a', 'z');
    ret += dis(gen);
  }
  return ret;
}

// O(n^2) solution
int CheckAns(const vector<string>& A, const vector<string>& Q) {
  unordered_set<string> dict;
  for (const string& s : Q) {
    dict.emplace(s);
  }

  Subarray ans = Subarray{0, static_cast<int>(A.size() - 1)};
  for (int l = 0; l < A.size(); ++l) {
    unordered_map<string, int> count;
    for (int r = l; r < A.size() && r - l < ans.end - ans.start; ++r) {
      if (dict.find(A[r]) != dict.end()) {
        ++count[A[r]];
      }
      if (count.size() == Q.size()) {
        if (r - l < ans.end - ans.start) {
          ans = {l, r};
        }
        break;
      }
    }
    count.clear();
  }
  return ans.end - ans.start;
}

void SimpleTestCase(const vector<string>& A, const vector<string>& dict,
                    int start, int finish) {
  auto res = FindSmallestSubarrayCoveringSet(A, {dict.begin(), dict.end()});
  cout << "res = " << res.start << " " << res.end << endl;
  assert(res.start == start && res.end == finish);
  string s;
  for (int i = 0; i < A.size(); ++i) {
    s += A[i];
    s += ' ';
  }
  istringstream sin(s);
  Subarray res2(FindSmallestSubarrayCoveringSubset(&sin, dict));
  cout << "res2 = " << res2.start << " " << res2.end << endl;
  assert(res2.start == start && res2.end == finish);
}

void SimpleTest() {
  vector<string> A = {"a", "b", "c", "b", "a", "d", "c",
                      "a", "e", "a", "a", "b", "e"};
  vector<string> dict = {"b", "c", "e"};
  SimpleTestCase(A, dict, 3, 8);
  dict = {"a", "c"};
  SimpleTestCase(A, dict, 6, 7);
  A = {"a", "b"};
  dict = {"a", "b"};
  SimpleTestCase(A, dict, 0, 1);
  A = {"a", "b"};
  dict = {"b", "a"};
  SimpleTestCase(A, dict, 0, 1);
}

int main(int argc, char* argv[]) {
  SimpleTest();
  default_random_engine gen((random_device())());
  for (int times = 0; times < 1000; ++times) {
    int n;
    vector<string> A;
    if (argc == 2) {
      n = atoi(argv[1]);
    } else {
      uniform_int_distribution<int> dis(1, 10000);
      n = dis(gen);
    }
    for (int i = 0; i < n; ++i) {
      uniform_int_distribution<int> dis(1, 10);
      A.emplace_back(RandString(dis(gen)));
    }
    /*
    for (int i = 0; i < A.size(); ++i) {
      cout << A[i] << ' ';
    }
    cout << endl;
    */
    unordered_set<string> dict;
    string s;
    for (int i = 0; i < A.size(); ++i) {
      dict.emplace(A[i]);
      s += A[i];
      s += ' ';
    }
    uniform_int_distribution<int> dis(1, dict.size());
    int m = dis(gen);
    vector<string> Q;
    for (auto it = dict.cbegin(); it != dict.cend(); ++it) {
      Q.emplace_back(*it);
      if (--m == 0) {
        break;
      }
    }

    /*
    for (int i = 0; i < Q.size(); ++i) {
      cout << Q[i] << ' ';
    }
    cout << endl;
    */
    unordered_set<string> input(Q.cbegin(), Q.cend());
    Subarray res(FindSmallestSubarrayCoveringSet(A, input));
    cout << res.start << ", " << res.end << endl;
    dict.clear();
    for (int i = 0; i < Q.size(); ++i) {
      dict.emplace(Q[i]);
    }
    for (int i = res.start; i <= res.end; ++i) {
      if (dict.find(A[i]) != dict.end()) {
        dict.erase(A[i]);
      }
    }
    assert(dict.empty() == true);
    istringstream sin(s);
    Subarray res2(FindSmallestSubarrayCoveringSubset(&sin, Q));
    cout << res2.start << ", " << res2.end << endl;
    dict.clear();
    for (int i = 0; i < Q.size(); ++i) {
      dict.emplace(Q[i]);
    }
    for (int i = res.start; i <= res.end; ++i) {
      if (dict.find(A[i]) != dict.end()) {
        dict.erase(A[i]);
      }
    }
    assert(dict.empty() == true);
    assert(res.end - res.start == res2.end - res2.start);
    assert(res.end - res.start == CheckAns(A, Q));
  }
  return 0;
}
